SmartDeviceLink/SDLEncryptionLifecycleManager.m

```diff
@@ -0,0 +1,241 @@
+}
+
+#pragma mark - SDLProtocolListener
+#pragma mark Encryption Start Service ACK
+
+- (void)handleProtocolStartServiceACKMessage:(SDLProtocolMessage *)startServiceACK {
+    switch (startServiceACK.header.serviceType) {
+        case SDLServiceTypeRPC: {
+            [self sdl_handleEncryptionStartServiceAck:startServiceACK];
+        } break;
+        default: break;
+    }
+}
+
+- (void)sdl_handleEncryptionStartServiceAck:(SDLProtocolMessage *)encryptionStartServiceAck {
+    if (encryptionStartServiceAck.header.encrypted) {
+        SDLLogD(@"Encryption service started");
+        [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateReady];
+    } else {
+        SDLLogD(@"Encryption service ACK received encryption = OFF");
+        [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+    }
+}
+
+#pragma mark Encryption Start Service NAK
+
+- (void)handleProtocolStartServiceNAKMessage:(SDLProtocolMessage *)startServiceNAK {
+    switch (startServiceNAK.header.serviceType) {
+        case SDLServiceTypeRPC: {
+            [self sdl_handleEncryptionStartServiceNAK:startServiceNAK];
+        } break;
+        default: break;
+    }
+}
+
+- (void)sdl_handleEncryptionStartServiceNAK:(SDLProtocolMessage *)audioStartServiceNak {
+    SDLLogW(@"Encryption service failed to start due to NAK");
+    [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+}
+
+#pragma mark Encryption End Service
+
+- (void)handleProtocolEndServiceACKMessage:(SDLProtocolMessage *)endServiceACK {
+    switch (endServiceACK.header.serviceType) {
+        case SDLServiceTypeRPC: {
+            SDLLogW(@"Encryption RPC service ended with end service ACK");
+            [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+        } break;
+        default: break;
+    }
+}
+
+- (void)handleProtocolEndServiceNAKMessage:(SDLProtocolMessage *)endServiceNAK {
+    switch (endServiceNAK.header.serviceType) {
+        case SDLServiceTypeRPC: {
+            SDLLogW(@"Encryption RPC service ended with end service NAK");
+            [self.encryptionStateMachine transitionToState:SDLEncryptionLifecycleManagerStateStopped];
+        } break;
+        default: break;
+    }
+}
+
+#pragma mark - SDL RPC Notification callbacks
+- (void)sdl_hmiStatusDidChange:(SDLRPCNotificationNotification *)notification {
+    if (![notification.notification isKindOfClass:[SDLOnHMIStatus class]]) {
+        return;
+    }
+    
+    SDLOnHMIStatus *hmiStatus = (SDLOnHMIStatus*)notification.notification;
+    self.hmiLevel = hmiStatus.hmiLevel;
+    
+    // if startWithProtocol has not been called yet, abort here
+    if (!self.protocol) { return; }
+    
+    if (!self.isEncryptionReady) {
+        [self sdl_startEncryptionService];
+    }
+}
+
+- (void)sdl_permissionsDidChange:(SDLRPCNotificationNotification *)notification {
```

Comment:
adding permission manager as oppose to observing this would require us to add a public method to permission manager that lets us see if any RPC contains the `requireEncryption` flag

Reply:
Why? You're not doing that right now in this manager, unless that's a bug?

Reply:
yea its a bug - according to the proposal we should wait for HMI to be full, background, limited and we should wait for onpermissionchange to check if at least one RPC needs protection

Reply:
another reason why this class should observe `OnPermissionChange` is in this scenario:
on the first `OnPermissionChange` no RPCs have a requiredEncryption flag so encryption service is not started.  At a later moment in time, Core sends another `OnPermissionChange`, this time with changed permissions where an RPC needs encryption so the service needs to be started.  If we rely on permission manager, then we would have to add new methods to the public header and also it will be hard to observe a change.